### DISCUSSION TITLE
tf.data.Dataset.shard error when distributing data

### DISCUSSION THREAD

**Discussion by gongel** [horovod/horovod#2774] [Q&A]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.15
3. Horovod version: 0.21.1
4. MPI version: 4.0.4
5. CUDA version: 10.2
6. NCCL version: 2.4.8
7. Python version: 3.7
8. OS and version: Debian
9. GCC version: 6.3

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
When using horovod to train BERT, I use tf.data.Dataset.shard to distribute data. But I ran into the following error:
```
[1,0]<stderr>:tensorflow.python.framework.errors_impl.UnknownError: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exception, you should see the exception in the log before the first shutdown message.
[1,0]<stderr>:   [[node DistributedAdamWeightDecayOptimizer_Allreduce/cond/HorovodAllreduce_gradients_AddN_48_0 (defined at /data00/home/maxiong/miniconda3/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:1748) ]]
```
My code is
 ```
        if is_training:
            d = tf.data.Dataset.from_tensor_slices(tf.constant(input_files))
            # distribute the data
            d = d.shard(hvd.size(), hvd.rank())
            d = d.shuffle(buffer_size=len(input_files))

            # `cycle_length` is the number of parallel files that get read.
            cycle_length = min(num_cpu_threads, len(input_files))

            # `sloppy` mode means that the interleaving is not exact. This adds
            # even more randomness to the training pipeline.
            d = d.apply(
                tf.contrib.data.parallel_interleave(
                    tf.data.TFRecordDataset,
                    sloppy=is_training,
                    cycle_length=cycle_length))
            d = d.repeat()
            # do mapping
            d = d.apply(
                tf.contrib.data.map_and_batch(
                    lambda record: tf.parse_single_example(record, name_to_features),
                    batch_size=batch_size,
                    num_parallel_batches=num_cpu_threads,
                    drop_remainder=True))```

**Comment by tgaddair** [comment]
Hey @gongel, what is likely happening here is that one of the workers is finishing before the others.  What you can do is call [tf.data.take()](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#take) to make sure all workers have the same number of batches.
